Introducing Solaris in OCI

anuthan opened the pull request:
This PR introduces Solaris application container configuration parameters to OCI.

Signed-off-by: Abhijeeth Nuthan <abhijeeth.nuthan@oracle.com>

vbatts commented:
nicely. On the whole, LGTM

wking commented on config-solaris.md:
```diff
@@ -0,0 +1,88 @@
+# Solaris Application Container Configuration
+
+Solaris application containers can be configured using the following properties, all of the below properties are the same as given in zonecfg(1M) man page, except milestone. The Solaris specification is entirely optional.
```
I'm not clear on “entirely optional”.  If you're aiming that at runtime authors (e.g. “authors of Linux runtimes can ignore this”), I don't think we need to call it out explicitly (all platform-specific specification can be ignored by runtime targeting other platforms).  If it's targeting bundle authors (e.g. “all settings in this section are optional, so even if you're writing a bundle targeting Solaris, you can leave this section out”), then I'd rather that get spelled out in more detail.  For example:

**`solaris`** *(object, optional)* Solaris-specific configuration.  Even configurations targeting Solaris MAY leave this field unset.

anuthan replied:
"entirely optional" means as you said, "solaris (object, optional) Solaris-specific configuration. Even configurations targeting Solaris MAY leave this field unset." 
Having said that, would it be more appropriate to the the above text in config.md as opposed to config-solaris.md and have a entry for each platform to say if the platform-specific section is required by the said platform.

wking replied:
On Fri, Apr 29, 2016 at 11:06:00AM -0700, Abhijeeth Nuthan wrote:
> Having said that, would it be more appropriate to the the above text
> in config.md as opposed to config-solaris.md and have a entry for
> each platform to say if the platform-specific section is required by
> the said platform.

Yeah, putting something more structured in [1] would be good.  It
looks like you're softening the existing SHOULD to a MAY.  I think
that's also the case for Linux, but I'm not sure, and having each
field listed with our usual:

**`{key}`** *({type}, {when-needed})* {description}

would make it more obvious.

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/config.md#platform-specific-configuration

anuthan replied:
Sounds like we need a separate PR to address the restructuring, if required. :)

wking replied:
On Fri, Apr 29, 2016 at 11:28:12AM -0700, Abhijeeth Nuthan wrote:
> Sounds like we need a separate PR to address the restructuring, if required. :) 

I just filed #414.

wking commented on config-solaris.md:
```diff
@@ -0,0 +1,88 @@
+## capped-cpu
+Sets a limit on the amount of CPU time that can be used by a container. The unit used translates to the percentage of a single CPU that can be used by all user threads in a container, expressed as a fraction (for example, .75) or a mixed number (whole number and fraction, for example, 1.25). An ncpu value of 1 means 100% of a CPU, a value of 1.25 means 125%, .75 mean 75%, and so forth. When proje ...
+
+* **`ncpus`** *(string, optional)*
```
Docs for **`capped-cpu`**?  For an example of this sort of nesting in the Linux docs, see [here][1].

[1]: https://github.com/opencontainers/runtime-spec/blob/v0.5.0/config-linux.md#memory

anuthan replied:
For every property , the documentation is zonecfg(1M) man page for that release of solaris, except "milestone" (which has no other docs apart from the explanation) . Same is mentioned at the top of the file. I did not see a need to call it out for each specific property.

wking replied:
On Fri, Apr 29, 2016 at 11:09:04AM -0700, Abhijeeth Nuthan wrote:
> > +* **`ncpus`** *(string, optional)*
>
> For every property, the documentation is zonecfg(1M) man page…

Right.  I just meant “why list ncpus, but not capped-cpu?”.

anuthan replied:
We want to keep the configuration same for all our flavors of virtualization. It's been this way for eons. So just continuing down that path made logical sense.

wking replied:
On Fri, Apr 29, 2016 at 11:22:23AM -0700, Abhijeeth Nuthan wrote:
> We want to keep the configuration same for all our flavors of
> virtualization…

I don't think I'm suggesting anything major, just:

* **`capped-cpu`** *(object, optional)* The following properties can be used to cap CPU usage:
  * **`ncpus`** *(string, optional)*

vishh replied:
If renaming is an option, why not just `cpus` :) ?

anuthan replied:
@wking : I'm assuming you are suggesting adding an (object, optional) for capped-cpu and capped-memory, etc. That would require rework for all the properties even in config-linux.md . I prefer not to do that in this PR.  

@vishh  : Sorry can't change names for legacy reasons.

wking replied:
On Fri, Apr 29, 2016 at 01:08:38PM -0700, Abhijeeth Nuthan wrote:
> > +* **`ncpus`** *(string, optional)*
>
> @wking : I'm assuming you are suggesting adding an (object,
> optional) for capped-cpu and capped-memory, etc. That would require
> rework for all the properties even in config-linux.md . I prefer not
> to do that in this PR.

I feel like we can work towards that goal in baby steps instead of
shifting everything at once.  But yeah, if it doesn't feel right to
you, we can certainly defer until someone is bothered enough to PR the
change.

wking commented on config.md:
```diff
@@ -108,16 +108,18 @@ For more information about SELinux, see  [Selinux documentation](http://selinuxp
 
 _Note: symbolic name for uid and gid, such as uname and gname respectively, are left to upper levels to derive (i.e. `/etc/passwd` parsing, NSS, etc)_
 
+_Note: For Solaris, uid and gid specify the uid and gid of the process inside the container and need not be same as in the host._
```
This is also true for Linux.  Instead of this note, I've filed #412 to make this explicit for both platforms.

wking commented on config.md:
```diff
@@ -608,6 +630,29 @@ Here is a full example `config.json` for reference.
         ],
         "mountLabel": "system_u:object_r:svirt_sandbox_file_t:s0:c715,c811"
     },
+    "solaris": {
```
We need a better way to handle this, because putting a `solaris` section in a config that has `linux` in `platform.os` may be confusing.

anuthan replied:
I was under the impression that we could have all platform specified as such in config.json . The runtime would only parse the required fields based on image and/or the platform section of the config.json.
We can refactor this however we see fit though. Need not necessarily be the way I suggested as above.

wking replied:
On Fri, Apr 29, 2016 at 11:19:26AM -0700, Abhijeeth Nuthan wrote:
> > +    "solaris": {
> 
> I was under the impression that we could have all platform specified
> as such in config.json . The runtime would only parse the required
> fields based on image and/or the platform section of the
> config.json.  We can refactor this however we see fit though. Need
> not necessarily be the way I suggested as above.

For previous multi-platform config discussion, see #73, #74, and #76.
The consensus was “this should be handled in a higher level” [1,2].

[1]: https://github.com/opencontainers/runtime-spec/issues/73#issuecomment-171449101
[2]: https://github.com/opencontainers/image-spec

wking commented on schema/schema-solaris.json:
```diff
@@ -0,0 +1,36 @@
+            },
+            "capped-cpu": {
+                "id": "https://opencontainers.org/schema/bundle/solaris/capped-cpu",
+                "$ref": "defs.json#/definitions/mapStringString"
```
This is accurate, but it could be more specific if you explicitly lists allowed keys.  That can happen with a [separate definition][1] or [inline][2].

[1]: https://github.com/opencontainers/runtime-spec/blob/e5e603f26fe8119b564e506d66e2b3426a68f25f/schema/schema-linux.json#L83-L86
[2]: https://github.com/opencontainers/runtime-spec/blob/e5e603f26fe8119b564e506d66e2b3426a68f25f/schema/schema-linux.json#L218-L243

anuthan replied:
This is just the initial PR, we will expand(more properties to come) and make it more structured in the future. For now, I think this would be fine.

wking replied:
On Fri, Apr 29, 2016 at 11:15:24AM -0700, Abhijeeth Nuthan wrote:
> > +            "capped-cpu": {
> > +                "id": "https://opencontainers.org/schema/bundle/solaris/capped-cpu",
> > +                "$ref": "defs.json#/definitions/mapStringString"
>
> This is just the initial PR, we will expand(more properties to come)
> and make it more structured in the future. For now, I think this
> would be fine.

Sounds good to me.

wking commented on specs-go/config.go:
```diff
@@ -25,6 +25,8 @@ type Spec struct {
 	// Linux is platform specific configuration for Linux based containers.
 	Linux Linux `json:"linux" platform:"linux"`
+	// Solaris is platform specific configuration for Solaris containers.
+	Solaris Solaris `json:"solaris" platform:"solaris"`
```
I'm not very familiar with Go structures, but I'd expect both `Linux` and `Solaris` should be pointers.

vishh commented on config-solaris.md:
````diff
@@ -0,0 +1,88 @@
+```
+
+## capped-cpu
+Sets a limit on the amount of CPU time that can be used by a container. The unit used translates to the percentage of a single CPU that can be used by all user threads in a container, expressed as a fraction (for example, .75) or a mixed number (whole number and fraction, for example, 1.25). An ncpu value of 1 means 100% of a CPU, a value of 1.25 means 125%, .75 mean 75%, and so forth. When proje ...
````
The difference between `ncpus` and `capped-cpu` is not clear by reading these comments. Can you explain based on an example with both values set, and/or provide links to additional documentation?

anuthan replied:
Both are actually one and the same , its just that structurally we expect them to be this way for legacy reasons and we can't really change the names or the way they are represented.  The value of ncpus is the value of capped-cpu. 

Example: Set container to use 80% of cpu.
"cappped-cpu" {
    "ncpus": "0.8"
}

Link to zonecfg(1M) manpage for Solaris 11.2: https://docs.oracle.com/cd/E36784_01/html/E36871/zonecfg-1m.html
You can find all the properties explained. It changes for each release of Solaris, hence we ask user to refer to zonecfg(1M) for additional information.

vishh replied:
Ok. Can you add the link to the PR itself?

On Fri, Apr 29, 2016 at 1:02 PM, Abhijeeth Nuthan <notifications@github.com>
wrote:

> In config-solaris.md
> <https://github.com/opencontainers/runtime-spec/pull/411#discussion_r61636723>
> :
>
> > +```json
> > +"limitpriv": "default"
> > +```
> > +
> > +## max-shm-memory
> > +The maximum amount of shared memory allowed for this application container. A scale (K, M, G, T) can be applied to the value for each of these numbers (for example, 1M is one megabyte).
> > +
> > +**`max-shm-memory`** *(string, optional)*
> > +
> > +### Example
> > +```json
> > +"max-shm-memory": "512m"
> > +```
> > +
> > +## capped-cpu
> > +Sets a limit on the amount of CPU time that can be used by a container. The unit used translates to the percentage of a single CPU that can be used by all user threads in a container, expressed as a fraction (for example, .75) or a mixed number (whole number and fraction, for example, 1.25). An ncpu value of 1 means 100% of a CPU, a value of 1.25 means 125%, .75 mean 75%, and so forth. When projects within a capped container have their own caps, the minimum value takes precedence.
>
> Both are actually one and the same , its just that structurally we expect
> them to be this way for legacy reasons and we can't really change the names
> or the way they are represented. The value of ncpus is the value of
> capped-cpu.
>
> Example: Set container to use 80% of cpu.
> "cappped-cpu" {
> "ncpus": "0.8"
> }
>
> Link to zonecfg(1M) manpage for Solaris 11.2:
> https://docs.oracle.com/cd/E36784_01/html/E36871/zonecfg-1m.html
> You can find all the properties explained. It changes for each release of
> Solaris, hence we ask user to refer to zonecfg(1M) for additional
> information.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/runtime-spec/pull/411/files/d57479d6f48fc50f584d566c26c9dc06da140946#r61636723>
>

wking replied:
On Fri, Apr 29, 2016 at 01:02:28PM -0700, Abhijeeth Nuthan wrote:
“It changes for each release of Solaris, hence we ask user to refer to
zonecfg(1M) for additional information.”

My intial thought on reading this was that you'll want some
target-Solaris-version information in the config.  But previous
discussion of that sort of thing (e.g. #39) ended up getting punted
over to image-spec.  Still, I'd suggest explicitly pointing out the
instability in your first config-solaris.md section.

anuthan replied:
@vishh : We try to keep the meaning of the properties same for all releases but the values that the properties can take are release specific. I do not want to put release specific info here in OCI. 
We can have certain keywords supported in one release but not the next/previous release. Giving a hypothetical example, "mac-address": "default", the "default" may be supported in Solaris 11.2 but need not be supported in future releases, so adding the link to Solaris11.2 would send the wrong information across. We advise the users to look at the man page for that specific release they are running on for valid properties values.  
Hope that clears things up.

@wking 
We had thought to introduce versioning to solaris object, but seemed unnecessary. Mainly because if we EOF or add properties we can up rev the ociVersion itself.

wking replied:
On Fri, Apr 29, 2016 at 01:42:57PM -0700, Abhijeeth Nuthan wrote:
“We can have certain keywords supported in one release but not the
next/previous release… We had thought to introduce versioning to
solaris object, but seemed unnecessary. Mainly because if we EOF or
add properties we can up rev the ociVersion itself.”

I am not looking forward to an runtime-spec major bump because Solaris
is dropping “default” support from “mac-address” in a future release
;).  In fact, I don't think we want to major bump the runtime-spec
because of any per-platform change.  I think re-opening per-feature
versioning (see #15 and the JSON-LD discussion in [1]) would be much
less invasive approaches to handling that sort of backwards
incompatibility.

[1]: https://github.com/opencontainers/runtime-spec/pull/371#issuecomment-209684002
     Subject: config.md: Rename ociVersion -> ociRuntimeVersion

anuthan replied:
I understand your concerns.
Let me just clarify what I said, might have confused you a bit.
1. The reason for not putting the a release specific link of a man page in this PR is to avoid unnecessary up-rev of versions.  That is, if we hypothetically remove support for "default" of "mac-address" , it would affect the spec.  No need to up-rev any versioning. 
2. If we remove a property/add property from the spec , it need not be even platform specific we still need to up-rev the spec.  I'm not saying we need to up-rev major/minor/patch but we still need to uprev. Whether it is platform specific version we need to up-rev or the ociVersion.  
Having said that, We should move this discussion of per platform specific versioning elsewhere as that is orthogonal to this PR.  
Thanks for raising your concerns. We should definitely discuss this further.

vishh commented:
@anuthan Overall LGTM. Is there an implementation of this Spec that users can play with today?

anuthan commented:
@vishh : The implementation of this spec is actively being worked on. I cannot comment on the timeline when this would be available. All I can say is coming soon to a Solaris store near you. :)

anuthan commented:
@vbatts and @vishh : Thanks for the LGTMs

crosbymichael commented on specs-go/config.go:
```diff
@@ -358,6 +360,51 @@ type Seccomp struct {
+	// Specify a name for the automatically created VNIC datalink.
+	Linkname string `json:"linkname,omitempty"`
+	// Specify the link over which the VNIC will be created.
+	Lowerlink string `json:"lower-link,omitempty"`
```
These need to be camel case and not include `-` to be consistent with the rest of the spec.

anuthan replied:
@crosbymichael  : As stated previously, the names of the properties which appear in the config.json , such as capped-cpu, capped-memory, lower-link, etc., have been part of Solaris virtualization for a long time. 
We prefer to keep the same name and structure to how it is defined in https://docs.oracle.com/cd/E36784_01/html/E36871/zonecfg-1m.html . 
Sorry if that ruins consistency.

crosbymichael commented on specs-go/config.go:
```diff
@@ -358,6 +360,51 @@ type Seccomp struct {
+	// Specification for automatic creation of network resources for this container.
+	Anet []Anet `json:"anet,omitempty"`
+	// Set limit on the amount of CPU time that can be used by container.
+	CappedCPU CappedCPU `json:"capped-cpu,omitempty"`
```
camel case

anuthan replied:
Same as previous comment.

crosbymichael commented:
Looks good overall but all your json fields need to be camel case and not include `-` because it makes it inconsistent with the spec as a whole.  

`max-shm-memory` to `maxShmMemory`

mrunalp commented on config-solaris.md:
```diff
@@ -0,0 +1,88 @@
+## Network
+
+### Automatic Network (anet)
+anet is specified as an array that is used to setup networking for Solaris application containers. The anet resource represents the automatic creation of a network resource for an application container. When such a container is started, a temporary VNIC(Virtual NIC) is automatically created for the container. The VNIC is deleted when the container is torn down.
```
One sentence per line.

anuthan replied:
If this is the only change you request, I will be happy to do it later today.

philips commented:
LGTM, I would like @anuthan to add a link or some explanatory text about the lifecycle of networking in Solaris is tied to a particular process as this is quite different than Linux where a majority of the expertise is coming from.

anuthan commented:
@philips : I have added text for life-cycle of networking and a pointer to our documentation, which is a man page of our administrative daemon. 

@crosbymichael : removed camel case.

crosbymichael commented:
LGTM

vbatts commented:
LGTM

anuthan commented:
@vbatts Thanks.

Also thanks to all who reviewed.
